### PR TITLE
Kafka Logger.

### LINKED ISSUE #1259: Add support to kafka publisher plugin
This plugin will work as a publisher which collects logs and publish to the external subscribers

### PR DESCRIPTION
**Pull request by Akayeshmantha**
Adding kafka plugin.

Issues resolved
Fix #1259

### CODE REVIEW

**Comment by Akayeshmantha** [comment]
@moonming @membphis can we review the PR ?

File: .travis.yml [hunk 1]
```diff
@@ -56,6 +56,13 @@ before_cache:
   - brew cleanup
 
 before_install:
+  - docker pull bitnami/zookeeper:3.6.0
```
Comment (membphis) [inline .travis.yml]: Not all CI environments support docker, such as OSX.

So we should complete the installation in a specific system script.

Here is an example: https://github.com/apache/incubator-apisix/blob/master/.travis/linux_openresty_runner.sh#L35

Comment (Akayeshmantha) [reply .travis.yml]: Sure.Since all tests passed which means we are currently skipping testcases run on osx?

Comment (membphis) [reply .travis.yml]: After this PR was merged, we started running test cases for the OSX system in each PR.

https://github.com/apache/incubator-apisix/pull/1314/files

Comment (Akayeshmantha) [reply .travis.yml]: @membphis 

I hope in the future we will be not running unit tests on 

OSNAME=linux_apisix_master_luarocks
OSNAME=linux_apisix_current_luarocks

Which I dont see currently we are running unit tests

Comment (Akayeshmantha) [reply .travis.yml]: @membphis  updated

File: .travis/osx_openresty_runner.sh [hunk 2]
```diff
@@ -33,6 +33,26 @@ before_install() {
     sudo cpanm --notest Test::Nginx >build.log 2>&1 || (cat build.log && exit 1)
     export_or_prefix
     luarocks install --lua-dir=${OPENRESTY_PREFIX}/luajit luacov-coveralls --local --tree=deps
+
+    # spin up kafka cluster for tests (1 zookeper and 1 kafka instance)
```
Comment (membphis) [inline .travis/osx_openresty_runner.sh]: we do not need to install the `zookeeper`, `kafka`.
only run the mini test cases(t/admin/*.t) for OSX system.

File: t/plugin/kafka-logger.t [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+=== TEST 5: access
+--- request
+GET /hello
+--- response_body
+hello world
+--- no_error_log
+[error]
+--- wait: 0.2
+
+=== TEST 6: error log
```
Comment (membphis) [inline t/plugin/kafka-logger.t]: use too `reindex` to check your test case style.

https://github.com/apache/incubator-apisix/blob/master/Contributing.md#check-code-style-and-test-case-style

File: lua/apisix/plugins/kafka-logger.lua [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    --Async producers will queue logs and push them when the buffer exceeds.
+    if conf.async then
+        broker_config["producer_type"] = "async"
+    end
+
+    local prod, err = producer:new(broker_list,broker_config)
+    if err then
+        core.log.error("failed to identify the broker specified", err)
+
```
Comment (membphis) [inline lua/apisix/plugins/kafka-logger.lua]: need to remove this blank line

File: lua/apisix/plugins/kafka-logger.lua [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        return
+    end
+
+    local ok, err = prod:send(conf.kafka_topic, conf.key, log_message)
+    if not ok then
+        core.log.error("failed to send data to Kafka topic", err)
+    end
+
```
Comment (membphis) [inline lua/apisix/plugins/kafka-logger.lua]: remove this blank line

**Comment by membphis** [comment]
@Akayeshmantha please rebase your branch with `master` branch

File: .travis/osx_openresty_runner.sh (hunk 2 above)
Comment (Akayeshmantha) [reply .travis/osx_openresty_runner.sh]: oh okays. I didnt know that

Comment (Akayeshmantha) [reply .travis/osx_openresty_runner.sh]: @membphis done.

File: lua/apisix/plugins/kafka-logger.lua (hunk 4 above)
Comment (Akayeshmantha) [reply lua/apisix/plugins/kafka-logger.lua]: done

File: t/plugin/kafka-logger.t (hunk 3 above)
Comment (Akayeshmantha) [reply t/plugin/kafka-logger.t]: done.

File: lua/apisix/plugins/kafka-logger.lua (hunk 5 above)
Comment (Akayeshmantha) [reply lua/apisix/plugins/kafka-logger.lua]: done.

**Comment by Akayeshmantha** [comment]
@membphis done

**Comment by membphis** [comment]
@Akayeshmantha It missing doc now.

The source code is fine.

**Comment by Akayeshmantha** [comment]
Added @membphis. And rebased with the master :smiley:

**Comment by membphis** [comment]
@Akayeshmantha many thx

**Pull request merged**